### PR TITLE
refactor find cluster version

### PR DESCRIPTION
**Pull request by myakove**
return only the cluster version and not full dict.

If user send `4.15` the version will be latest from stream (stable, nightly, candidate, etc)
If user send `4.15.7` the version will be the user version if the version is valid for the platform and the stream

### CODE REVIEW

**Comment by redhat-qe-bot2** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest build-container`: Retest build-container
 * `/retest python-module-install`: Retest python-module-install

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: openshift_cli_installer/libs/clusters/ocm_cluster.py [hunk 1]
```diff
@@ -42,18 +44,29 @@ def _set_expiration_time(self):
...
-            Versions(client=self.ocm_client).get(channel_group=self.cluster_info["channel-group"])
-        )
+        updated_versions_dict: Dict[str, Dict[str, List[str]]] = {}
+        for cannel, versions in (
+            Versions(client=self.ocm_client).get(channel_group=self.cluster_info["channel-group"]).items()
+        ):
+            updated_versions_dict[cannel] = {}
+            for version in versions:
+                _version_key = re.findall(r"^\d+.\d+", version)[0]
+                updated_versions_dict[cannel].setdefault(_version_key, []).append(version)
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/ocm_cluster.py:54]:
```suggestion
                updated_versions_dict.setdefault(cannel, {}).setdefault(_version_key, []).append(version)
```

File: openshift_cli_installer/libs/clusters/ocm_cluster.py [hunk 2]
```diff
@@ -42,18 +44,29 @@ def _set_expiration_time(self):
...
 
     def get_osd_versions(self):
-        self.osd_base_available_versions_dict.update(
-            Versions(client=self.ocm_client).get(channel_group=self.cluster_info["channel-group"])
-        )
+        updated_versions_dict: Dict[str, Dict[str, List[str]]] = {}
+        for cannel, versions in (
+            Versions(client=self.ocm_client).get(channel_group=self.cluster_info["channel-group"]).items()
+        ):
+            updated_versions_dict[cannel] = {}
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/ocm_cluster.py:51]: missing typing

File: openshift_cli_installer/libs/clusters/ocm_cluster.py [hunk 3]
```diff
@@ -42,18 +44,29 @@ def _set_expiration_time(self):
             )
 
     def get_osd_versions(self):
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/ocm_cluster.py:46]: we should add caching here, right?

File: openshift_cli_installer/libs/clusters/ocm_cluster.py [hunk 4]
```diff
@@ -42,18 +44,29 @@ def _set_expiration_time(self):
...
+            Versions(client=self.ocm_client).get(channel_group=self.cluster_info["channel-group"]).items()
+        ):
+            updated_versions_dict[cannel] = {}
+            for version in versions:
+                _version_key = re.findall(r"^\d+.\d+", version)[0]
+                updated_versions_dict[cannel].setdefault(_version_key, []).append(version)
+
+        self.osd_base_available_versions_dict.update(updated_versions_dict)
 
     def get_rosa_versions(self):
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/ocm_cluster.py:58]: ditto?

File: openshift_cli_installer/utils/cluster_versions.py [hunk 5]
```diff
@@ -21,116 +19,39 @@
...
-def filter_versions(wanted_version, base_versions_dict, platform, stream):
-    versions_dict = {}
-    version_key = get_split_version(version=wanted_version)
-    versions_dict[stream] = {version_key: {"versions": set(), "latest": ""}}
+def get_cluster_version_to_install(
+    wanted_version: str, base_versions_dict: Dict, platform: str, stream: str, log_prefix: str, cluster_name: str
+) -> str:
+    wanted_version_len = len(wanted_version.split("."))
+    if wanted_version_len < 2:
+        LOGGER.error(f"{cluster_name}: Version must be at least x.y (4.3), got {wanted_version}")
```
Comment (rnetser) [inline openshift_cli_installer/utils/cluster_versions.py:27]:
```suggestion
        LOGGER.error(f"{log_prefix}: Version must be at least x.y (4.3), got {wanted_version}")
```

File: openshift_cli_installer/utils/cluster_versions.py [hunk 6]
```diff
@@ -21,116 +19,39 @@
...
+                match = _match[0]
+                break
 
-        if all_semver_versions:
-            max_version = str(max([ver for ver in all_semver_versions]))
-            versions_dict[stream][version_key]["latest"] = f"{max_version}"
-
-    if not versions_dict[stream][version_key]["versions"]:
+    if not match:
         LOGGER.error(f"Cluster version {wanted_version} not found for stream {stream}")
```
Comment (rnetser) [inline openshift_cli_installer/utils/cluster_versions.py:50]:
```suggestion
        LOGGER.error(f"{log_prefix}Cluster version {wanted_version} not found for stream {stream}")
```

File: openshift_cli_installer/tests/cluster_version/test_osd_rosa_versions.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
+import click
+import pytest
+
+from openshift_cli_installer.tests.cluster_version.rosa_osd_base_versions import ROSA_OSD_BASE_VERSIONS
+from openshift_cli_installer.utils.cluster_versions import (
+    get_cluster_version_to_install,
+)
+
+PARAMETRIZE_TESTS = [
+    ([{"version": "4.15", "stream": "stable", "expected": "4.15.8"}]),
```
Comment (rnetser) [inline openshift_cli_installer/tests/cluster_version/test_osd_rosa_versions.py:10]: we should add a check in user input that IP get the right streams and rosa etc get the right channel-group

File: openshift_cli_installer/tests/cluster_version/test_osd_rosa_versions.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            "version": "4.16.0-0.nightly-2024-04-16-195622",
+            "stream": "nightly",
+            "expected": "4.16.0-0.nightly-2024-04-16-195622",
+        }
+    ]),
+    ([{"version": "4.15.8", "stream": "stable", "expected": "4.15.8"}]),
+    ([{"version": "4.15", "stream": "candidate", "expected": "4.15.9"}]),
+]
+
+PARAMETRIZE_NEGATIVE_TESTS = [
```
Comment (rnetser) [inline openshift_cli_installer/tests/cluster_version/test_osd_rosa_versions.py:30]: same as ipi negative dict

**Review by rnetser — CHANGES_REQUESTED** [review]

File: openshift_cli_installer/libs/clusters/ocm_cluster.py [hunk 9]
```diff
@@ -42,18 +44,29 @@ def _set_expiration_time(self):
             )
 
     def get_osd_versions(self):
-        self.osd_base_available_versions_dict.update(
-            Versions(client=self.ocm_client).get(channel_group=self.cluster_info["channel-group"])
-        )
+        updated_versions_dict: Dict[str, Dict[str, List[str]]] = {}
+        for cannel, versions in (
```
Comment (omrirh) [inline openshift_cli_installer/libs/clusters/ocm_cluster.py:48]:
```suggestion
        for channel, versions in (
```

File: openshift_cli_installer/libs/clusters/ocm_cluster.py (hunk 2 above)
Comment (myakove) [reply openshift_cli_installer/libs/clusters/ocm_cluster.py]: No, you can't add typing to '{}['something']'

File: openshift_cli_installer/libs/clusters/ocm_cluster.py (hunk 1 above)
Comment (myakove) [reply openshift_cli_installer/libs/clusters/ocm_cluster.py]: No need, we have `updated_versions_dict[channel] = {}`

**Review by rnetser — APPROVED** [review]

**Comment by msi-b1** [comment]
New container for quay.io/redhat_msi/openshift-cli-installer:latest published

**Pull request merged**